### PR TITLE
Refactor - Modularization of javalib modules

### PR DESCRIPTION
**Pull request by WojciechMazur**
This PR finalizes work done in PR #1672. 
The mentioned PR was rebased on the current master.  Also in order to allow compilation of javalib modules dependent on `javalibcore` there was introduced conditional including of `.class` files in resulting JAR based on filters in `class-whitelist` files. 
Current filters only include files that should not exist in any JDK or are ScalaNative specific. 
This also creates an opportunity for further modularization of javalib. 

Additionally, before it is merged I would like to discuss the structure of the new `javalib` modules. Currently, flat structure is used (similarly as in Scala.js). However, if we would like to further extract modules from java core (ex. security, time, etc.) it might be better to keep them all in common javalib directory. 
New structure may be presented as below:
```
| javalib:
|    |--- core
|    |--- net
|    |--- file
|    |--- zip
| <other-modules>
```

### CODE REVIEW

**Comment by ekrich** [comment]
I don't know if anyone has reached out to @muxanick but he was always way in front of the curve with Scala Native and we could never quite get to where he wanted to be. Finally, we are moving in that direction and quite a bit quicker now.

**Comment by LeeTibbert** [comment]
There are currently 8 or 9 pending PRs affecting javalib.  That is roughly 10% of the backlog.
Those 8 or 9  not count the whole whack of JavaDefaultMethod PRs I am in the midst of doing.
Some of those are more difficult than they appear.

Such a large restructuring will basically invalidate all of that work.   It would be nice if 
those PRs could at least be examined and triaged.   They are value in hand, waiting to
ship.

**Comment by ekrich** [comment]
@LeeTibbert As one that also begged mercy, I totally understand. I am sure the team will try and wait for those to land before pushing this through.

**Comment by LeeTibbert** [comment]
I am aware that a 400+ file PR has a short expiry date, having been there myself.

I  can adapt my current work, especially now that the nscplugin change is in, so that should not
hold this PR up.

Some of the 8 or 9 old ones, such as  the File nanosecond time one, are probably too complex, causing too long a wait.
I think there are a number of others which are as easy as anything ever is; low hanging fruit which will be abandoned.

**Comment by ekrich** [comment]
I think that conditional compilation is probably good for the basic core like we have so far but much beyond that and we end up having to do too much abstraction work in C. Either a POSIX layer for windows or some sort of other poor compromise.

What are some thoughts on how we might approach a Windows version of `java.io` for example?

Would we have some type of overlays or separate projects that build conditionally?

The initial thoughts about this approach were that a Windows port could work side by side on its own timeline and not affect the stability on the mainline.

I am really not sure about the Java lib layout but it probably depends on the overall strategy.

https://github.com/scala-native/scala-native/issues/113

**Comment by WojciechMazur** [comment]
@LeeTibbert  It's not a problem with waiting with other PR to be merged. Most of the changes are just moving files and don't cause any problems when rebasing.

**Comment by WojciechMazur** [comment]
Actually merging this PR creates multiple problems due to dependencies in different versions of Scala. 
For. eg. since 2.12 `scala.compat.Platfrom`  uses Properties (in javalib `core`) in order to get `lineSeperator`. This however needs `java.util.jar.Attributes$Name` which is placed in `zip`. Also `j.u.Runtime` is placed.
Also `j.u.Runtime` currently is tightly dependant on `UnixProcess` and `PipeIO` using `posix`. This means we always would need to import this package anyway. 

To summarize: Modularization might be a good idea,  but I don't think the current approach is correct. Currently, it's too tightly coupled. To allow modularization we should develop a model where we could define a reference interface or dummy implementation that would be implemented/overriden for each target system.

**Comment by LeeTibbert** [comment]
Can you/we create an (current) Issue for discussion & design? Would that
be a more appropriate venue than a PR review comment?

**Comment by WojciechMazur** [comment]
Modularization of javalib introduces multiple hard to solve issues, especially with circular dependencies, or package-private scopes not accessible from other modules. Main purpose of modularization was the easier implementation of javalib across different platforms, it was already fulfilled using link-time conditions.
